src/schemas/tdc_time/TdcTime.cpp

```diff
@@ -0,0 +1,92 @@
+#include "TdcTime.h"
+#include "../../EpicsPVUpdate.h"
+#include "../../RangeSet.h"
+#include "../../logger.h"
+#include <algorithm>
+#include <atomic>
+#include <mutex>
+#include <pv/nt.h>
+#include <pv/ntndarray.h>
+#include <pv/ntndarrayAttribute.h>
+#include <pv/ntutils.h>
+#include <pv/pvEnumerated.h>
+#include <string>
+#include <tdct_timestamps_generated.h>
+
+namespace TdcTime {
+
+namespace pvNT = epics::nt;
+namespace pv = epics::pvData;
+
+template <typename pvScalarArrType>
+std::vector<std::uint64_t> CalcTimestamps(pv::PVFieldPtr const &scalarArray) {
+  auto pvValues = dynamic_cast<pvScalarArrType *>(scalarArray.get());
+  size_t NrOfElements = pvValues->getLength();
+  if (NrOfElements % 2 != 0) {
+    throw std::runtime_error("Size of pvAccess array is not a power of two.");
+  }
+  const auto pvArrPtr = pvValues->view().data();
+  std::vector<std::uint64_t> RetVector;
+  for (size_t i = 0; i < NrOfElements; i += 2) {
+    RetVector.push_back(pvArrPtr[i] * 1000000000L + pvArrPtr[i + 1]);
+  }
+  return RetVector;
+}
+
+std::unique_ptr<FlatBufs::FlatbufferMessage>
+Converter::create(FlatBufs::EpicsPVUpdate const &PvData) {
+  auto pvUpdateStruct = PvData.epics_pvstr;
+  std::string pvStructType = pvUpdateStruct->getField()->getID();
+  if (pvStructType != "epics:nt/NTScalarArray:1.0") {
+    getLogger()->critical("PV is not of expected type for chopper TDC.");
+    return {};
+  }
+  pvNT::NTScalarArrayPtr ntScalarData =
+      pvNT::NTScalarArray::wrap(pvUpdateStruct);
+  auto scalarArrPtr =
+      dynamic_cast<pv::PVScalarArray *>(ntScalarData->getValue().get());
+  auto ElementType = scalarArrPtr->getScalarArray()->getElementType();
+  if (ElementType != pv::ScalarType::pvInt) {
+    getLogger()->error("Array elements are not of expected type.");
+    return {};
+  } else {
+    auto ConvertField = ntScalarData->getValue();
+    try {
+      auto NewTimestamps = CalcTimestamps<pv::PVIntArray>(ConvertField);
+      if (NewTimestamps.empty()) {
+        return {};
+      }
+      return generateFlatbufferFromData(PvData.channel, NewTimestamps);
+    } catch (std::runtime_error &E) {
+      getLogger()->critical("Unable to convert pv-array into timestamps: {}",
+                            E.what());
+      return {};
+    }
+  }
+  return {};
+}
+
+std::unique_ptr<FlatBufs::FlatbufferMessage>
+generateFlatbufferFromData(std::string const &Name,
```

Comment:
`generateFlatbufferFromData` is tested via `Converter::create`, which is fine, but means we may as well reduce work for the compiler and put it in the anonymous namespace for this file.

Reply:
A quick google search yields no results which indicates that anonymous namespaces reduces compile time.

Reply:
Not sure what I was thinking about compile time... but from C++11 things in the anonymous namespace are completely local to that translation unit. I guess this means fewer symbols lying around for the linker to deal with, no idea if that is significant. I've seen some people claim it can even mean runtime performance improvement though, because the compiler is more likely to be able to inline things: https://stackoverflow.com/a/25565298/5887927

Reply:
In general though, I see it as part of the "put stuff in the smallest scope you can" idea.

Reply:
Fair enough, the last argument I can agree with.